packages/backend-defaults/src/entrypoints/urlReader/lib/BitbucketCloudUrlReader.ts

```diff
@@ -69,6 +73,8 @@ export class BitbucketCloudUrlReader implements UrlReaderService {
         `Bitbucket Cloud integration for '${host}' has configured a username but is missing a required appPassword.`,
       );
     }
+
+    this.fetch = BitbucketFetchService.get(this.integration.config);
```

Comment:
Note the [integration-aws-node package](https://github.com/backstage/backstage/tree/526d5f1101a299cd0a9b0fe43c22355ef4a6bec3/packages/integration-aws-node). That integration got to a point where they realized that they couldn't stay within the confines of a polymorphic (web + node at the same time) package so their integration-specific code had to go in a package of its own. Crucially, _that package has a different role_ in its package.json, namely [node-library](https://github.com/backstage/backstage/blob/526d5f1101a299cd0a9b0fe43c22355ef4a6bec3/packages/integration-aws-node/package.json#L6) instead of common-library like the polymorpic ones. This is important because the build setup for those are entirely different and they are permitted to import node packages like node-fetch. And it let the regular integrations package stay slim.

The code that builds your special fetch function has to live in a node-library role package if it wants to be based on node-fetch. Unfortunately, bitbucket-cloud-common is not of that role. You have two options:

- First option: Rewrite your fetcher to be polymorphic, base it on only the cross-fetch library instead of node-fetch, and move its code into bitbucket-cloud-common. This would be kind of nice, but I expect that there might turn out to be some small differences between cross-fetch and node-fetch, making typescript upset when using that fetcher in your node packages.
- Second option: Create an integration-bitbucket-node package, make it of the node-library role, and move your implementation there. Then of course it won't be polymorphic.

The thing is, I think bitbucket-cloud-common should never have been a -common package in the first place. It is only used by backend packages, so making it polymorphic just made everything harder for no reason. I vote doing option 2, and (later down the line) deprecating bitbucket-cloud-common and moving all of its guts into the new integration-bitbucket-node package too. Don't do that last bit now though, we have enough complexity to handle anyway.

Adding one more thing. You want to be able to pass in the custom fetch function in the constructor (in the deps argument), instead of constructing it here. Then tests and alternative callers can pass in what they want. You may want to create the actual fetcher inside the factory above instead of the constructor.